File: SnakeAndLadder.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace BFSnDFS
{
    class SnakeAndLaddersLC
    {
        //TC: O(n^2)
        //SC: O(n^2)
        public int SnakesAndLadders(int[][] board)
        {
            if (board == null || board.Length == 0) return - 1;
            int n = board.Length;
            int[] moves = new int[n * n];
            int i = n - 1, j = 0;
            int idx = 0;
            int even = 0;
            while (i >= 0 && j >= 0)
            {
                if (board[i][j] == -1)
                {
                    moves[idx] = board[i][j];
                }
                else
                {
                    moves[idx] = board[i][j] - 1;
                }
                idx++;
                if (even % 2 == 0)
                {
                    j++;
                    if (j == n)
                    {
                        j--;
                        even++;
                        i--;
                    }
                }
                else
                {
                    j--;
                    if (j < 0)
                    {
                        j++;
                        even++;
                        i--;
                    }
                }
            }
            Queue<int> q = new Queue<int>();
            int moveCount = 0;
            q.Enqueue(0);
            moves[0] = -2;
            while (q.Count != 0)
            {
                int size = q.Count;
                for (int k = 0; k < size; k++)
                {
                    int curr = q.Dequeue();
                    if (curr == n * n - 1) return moveCount;
                    for (int l = 1; l <= 6; l++)
                    {
                        int pos = curr + l;
                        if (moves[pos] != -2)
                        {
                            if (moves[pos] == -1)
                            {
                                moves[pos] = -2;
                                q.Enqueue(pos);
                            }
                            else
                            {
                                q.Enqueue(moves[pos]);
                                moves[pos] = -2;
                            }
                        }
                    }
                }
                moveCount++;
            }
            return -1;
        }
    }
}

File: snakeAndLadder.py
from collections import deque

#Time Complexity : O(N^2)
#Space Complexity : O(N^2)

class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        if not board:
            return -1

        result = 0
        n = len(board)
        arrBoard = [-1] * (n * n)
        idx = 0
        i = n - 1
        j = 0
        even = False
        while idx < n * n:
            if board[i][j] >= 1:
                arrBoard[idx] = board[i][j] - 1

            idx += 1
            # new r and c
            if not even:
                j += 1
                if j == n:
                    i -= 1
                    j -= 1
                    even = True
            else:
                j -= 1
                if j < 0:
                    i -= 1
                    j += 1
                    even = False
        # print(arrBoard)

        que = deque([0])
        arrBoard[0] = -2

        while que:
            size = len(que)
            for i in range(size):
                curr = que.popleft()
                for i in range(1, 7):
                    child = curr + i
                    # if child > (n*n):
                    #     break
                    if arrBoard[child] != -2:
                        if arrBoard[child] == -1:
                            if child == n * n - 1:
                                return result + 1
                            que.append(child)

                        else:
                            if arrBoard[child] == n * n - 1:
                                return result + 1
                            que.append(arrBoard[child])
                        arrBoard[child] = -2
            result += 1
        return -1